caddy.go: Check whether @id is unique(#6991)

Chise1 opened the pull request:
Fixes: https://github.com/caddyserver/caddy/issues/6991

Check whether @id is unique.If it is not unique,rollback caddy config json.
I'm not sure if it can work properly under caddyfile.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/caddyserver/caddy?pullRequest=7002) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/caddyserver/caddy?pullRequest=7002) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/caddyserver/caddy?pullRequest=7002) it.</sub>

codehakase commented on caddy.go:
```diff
@@ -224,6 +224,10 @@ func changeConfig(method, path string, input []byte, ifMatchHeader string, force
 	idx := make(map[string]string)
 	err = indexConfigObjects(rawCfg[rawConfigKey], "/"+rawConfigKey, idx)
 	if err != nil {
+		err2 := rollbackRawCfg()
+		if err2 != nil {
+			err = errors.Join(err, err2)
+		}
```
Trying to parse the old config here without checking the length of `rawCfgJSON` isn't a recommended approach. Additionally, this is a potential duplicate of the call in `Ln246`

codehakase commented on caddy.go:
```diff
@@ -261,13 +263,28 @@ func changeConfig(method, path string, input []byte, ifMatchHeader string, force
 
+func rollbackRawCfg() error {
+	var oldCfg any
+	err := json.Unmarshal(rawCfgJSON, &oldCfg)
+	rawCfg[rawConfigKey] = oldCfg
```
obs: Potential nil assignment to `rawCfg`. Check for error before assigning to the map

Chise1 commented:
I added length detection and fixed a stupid problem. From the perspective of code analysis, "err" will not be repeated.

Changes of rewrite, involves [nginx-adapter-case](:https://github.com/caddyserver/nginx-adapter/issues/35)

mohammed90 commented:
> Changes of rewrite, involves nginx-adapter-case

Please keep the PR topical and focused. Split these changes into a separate PR.

Chise1 commented:
It seems that there are many problems now. I need to think carefully again about how to modify it. Please ignore my pr for the moment

mholt commented:
I feel like a better approach would be to check the ID as it's being read in and we're creating the mapping of IDs to their full config paths. If we return an error the rest should be handled automatically (no need to roll back, etc)

Chise1 commented:
> I feel like a better approach would be to check the ID as it's being read in and we're creating the mapping of IDs to their full config paths. If we return an error the rest should be handled automatically (no need to roll back, etc)

If we do as you said, the content that needs to be modified is rather complicated.  This is beyond my ability.  I'm worried that the modification will introduce bugs.

Chise1 commented:
add one test.

francislavoie commented:
Since this fell inactive, I took it upon myself to try and reimplement this.

The tests were good, thanks! It was a good starting point.

I pointed Gemini 3.1 Pro + Github Copilot at this to take a shot at doing it more cleanly. I rebased and pushed my version on top. The result is quite a lot simpler. The new functions weren't really necessary.

mholt reviewed (approved):
Thanks Francis. Glad to merge this in.
